File: data-inspector-scala/src/main/scala/com/example/datainspector/service/DataInspectorService.scala
/*
 * Copyright (c) 2026 Data Inspector Contributors
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.example.datainspector.service

import com.example.datainspector.model.{DataSourceInfo, QueryResult}
import com.example.datainspector.spi.DataSourceProvider
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.stereotype.Service

import scala.jdk.CollectionConverters._

/**
 * Core service that aggregates all data sources and handles queries
 */
@Service
class DataInspectorService(providers: java.util.List[DataSourceProvider]) {

  private val logger: Logger = LoggerFactory.getLogger(classOf[DataInspectorService])
  private val providerList: List[DataSourceProvider] = providers.asScala.toList

  logger.info(s"Data Inspector initialized with ${providerList.size} provider(s)")

  /**
   * Get all available data sources from all providers
   */
  def getAllDataSources(): java.util.List[DataSourceInfo] = {
    providerList.flatMap(_.discoverDataSources()).asJava
  }

  /**
   * Query a specific data source
   */
  def query(dataSourceId: String, filters: java.util.Map[String, AnyRef], limit: Int, offset: Int): QueryResult = {
    val scalaFilters = Option(filters).map(_.asScala.toMap).getOrElse(Map.empty[String, AnyRef])

    val provider = providerList
      .find(_.supports(dataSourceId))
      .getOrElse(throw new IllegalArgumentException(s"No provider found for data source: $dataSourceId"))

    provider.query(dataSourceId, scalaFilters, limit, offset)
  }

  /**
   * Get a specific data source by ID
   */
  def getDataSource(dataSourceId: String): DataSourceInfo = {
    providerList
      .flatMap(_.discoverDataSources())
      .find(_.id == dataSourceId)
      .getOrElse(throw new IllegalArgumentException(s"Data source not found: $dataSourceId"))
  }
}

File: data-inspector/src/main/java/io/github/kanxiao/spring/inspector/service/DataInspectorService.java
/*
 * Copyright (c) 2026 Data Inspector Contributors
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package io.github.kanxiao.spring.inspector.service;

import io.github.kanxiao.spring.inspector.model.DataSourceInfo;
import io.github.kanxiao.spring.inspector.model.QueryResult;
import io.github.kanxiao.spring.inspector.spi.DataSourceProvider;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Core service that aggregates all data sources and handles queries
 */
@Slf4j
public class DataInspectorService {

    private final List<DataSourceProvider> providers;

    public DataInspectorService(List<DataSourceProvider> providers) {
        this.providers = providers;
        log.info("Data Inspector initialized with {} provider(s)", providers.size());
    }

    /**
     * Get all available data sources from all providers
     */
    public List<DataSourceInfo> getAllDataSources() {
        return providers.stream()
                .flatMap(provider -> provider.discoverDataSources().stream())
                .collect(Collectors.toList());
    }

    /**
     * Query a specific data source
     */
    public QueryResult query(String dataSourceId, Map<String, Object> filters, int limit, int offset) {
        DataSourceProvider provider = providers.stream()
                .filter(p -> p.supports(dataSourceId))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("No provider found for data source: " + dataSourceId));

        return provider.query(dataSourceId, filters, limit, offset);
    }

    public Object executeAction(String dataSourceId, String action, Map<String, Object> params) {
        DataSourceProvider provider = providers.stream()
                .filter(p -> p.supports(dataSourceId))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("No provider found for data source: " + dataSourceId));

        return provider.executeAction(dataSourceId, action, params);
    }

    /**
     * Get a specific data source by ID
     */
    public DataSourceInfo getDataSource(String dataSourceId) {
        return getAllDataSources().stream()
                .filter(ds -> ds.getId().equals(dataSourceId))
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("Data source not found: " + dataSourceId));
    }
}
